Computer Science (both courses): Improve space complexity content (#30463)
TheOdinProject/curriculum: javascript/computer_science/space_complexity.md

- What do we mean by space complexity.
- Why is it important to consider how our algorithm uses memory space.
- How do we measure space complexity.

> - What is a situation where auxiliary space analysis is useful.

Reviewer: Some of these are worded as questions but with "fixed" punctuation because of TOP009 but at the same time, [LOs should be a list of topics covered in the lesson](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md#lesson-layout), so something like the below would be preferable (and match the default section content "This section contains a general overview of topics that you will learn in this lesson.").

```suggestion
- What space complexity is.
- The importance of space complexity.
- How to measure space complexity.
- Auxiliary space analysis.
```

}
```

> When a data structure is passed in as the argument, especially for languages that pass arrays by reference rather than value, it can be a bit unclear if that method considers the space used by that data structure when calculating its space complexity. We could still say this algorithm requires O(N) space, factoring in the size of the input. Or we could say that the creation of the input array belongs to whatever external routine allocated it before calling this function. This would mean that this algorithm takes up O(1) space, as all other variables are constant. This latter approach is known an "auxiliary space analysis." Using this approach, we only regard the *extra* space that an algorithm takes up, meaning the input isn't factored in. This can be situationally useful for describing algorithms that use extra space differently. Let's look at a quick example:

Reviewer: Feels like a p-break here would be easier to follow?
```suggestion
When a data structure is passed in as the argument, especially for languages that pass arrays by reference rather than value, it can be a bit unclear if that method considers the space used by that data structure when calculating its space complexity. We could still say this algorithm requires O(N) space, factoring in the size of the input. Or we could say that the creation of the input array belongs to whatever external routine allocated it before calling this function. This would mean that this algorithm takes up O(1) space, as all other variables are constant. This latter approach is known an "auxiliary space analysis."

Using this approach, we only regard the *extra* space that an algorithm takes up, meaning the input isn't factored in. This can be situationally useful for describing algorithms that use extra space differently. Let's look at a quick example:
```

```javascript
function squareNumsInPlace(arr) {
  for (let i = 0; i < arr; i++) {
    arr[i] = arr[i] * arr[i];
  }

  return arr;
}

function squareNumsNewArr(arr) {
  const squaredNums = [];
  arr.forEach((number) => {
    squaredNums.push(number * number);
  })

  return squaredNums;
}
```

Here we have two functions that take in an array of numbers and return an array where each number has been squared. The first function directly changes the input array, while the second function builds a new array to `push()` the squared numbers into. Using traditional space analysis, both functions would take up O(N) space. But for this situation, that doesn't tell us much about the (very clear) differences between how these two algorithms use memory. It'd be better to use auxiliary space analysis and say that the top function uses O(1) extra space while the bottom one uses O(N).

#### Other complexities

As we've stated, many data structures share O(N) space complexity, and therefore you won't write many algorithms with a space complexity that differs.

> You do find some recursive functions that may have a different space complexity and some sorting algorithms. You normally won't have much reason to consider anything else though.

Reviewer: Let me know how the wording suggestion flows
```suggestion
You do find some recursive functions and sorting algorithms with a different space complexity, though you normally won't have much reason to consider anything else.
```

}
```

When a data structure is passed in as the argument, especially for languages that pass arrays by reference rather than value, it can be a bit unclear if that method considers the space used by that data structure when calculating its space complexity. We could still say this algorithm requires O(N) space, factoring in the size of the input. Or we could say that the creation of the input array belongs to whatever external routine allocated it before calling this function. This would mean that this algorithm takes up O(1) space, as all other variables are constant. This latter approach is known an "auxiliary space analysis."

Using this approach, we only regard the *extra* space that an algorithm takes up, meaning the input isn't factored in. This can be situationally useful for describing algorithms that use extra space differently. Let's look at a quick example:

```javascript
function squareNumsInPlace(arr) {
  for (let i = 0; i < arr; i++) {
    arr[i] = arr[i] * arr[i];
  }

  return arr;
}

function squareNumsNewArr(arr) {
  const squaredNums = [];
  arr.forEach((number) => {
    squaredNums.push(number * number);
  });

  return squaredNums;

> }

Author: It is to be extra clear about the extra array being created, and it's aligned with how the lesson previously wrote these examples.

That said, I do think learners are familiar enough with `map` that it should work, and it's definitely more idiomatic (in both languages).

I still might include a comment that draws direct attention to `map` creating a *new* array. Does that seem reasonable?

> }

Reviewer: I dig it
